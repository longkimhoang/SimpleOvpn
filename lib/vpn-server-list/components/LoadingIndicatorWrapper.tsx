import React from 'react';
import {StyleSheet} from 'react-native';
import {ActivityIndicator} from 'react-native-paper';
import Animated, {useAnimatedStyle, withTiming} from 'react-native-reanimated';

export interface LoadingIndicatorWrapperProps {
  isFetching: boolean;
}

function LoadingIndicatorWrapper({isFetching}: LoadingIndicatorWrapperProps) {
  const animatedStyle = useAnimatedStyle(() => {
    return {
      opacity: withTiming(isFetching ? 1 : 0),
    };
  }, [isFetching]);

  return (
    <Animated.View
      pointerEvents="none"
      style={[styles.loadingIndicator, animatedStyle]}
      testID="vpn-server-list.loading-indicator.container">
      <ActivityIndicator animating={isFetching} hidesWhenStopped={false} />
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  loadingIndicator: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
  },
});

export default LoadingIndicatorWrapper;
