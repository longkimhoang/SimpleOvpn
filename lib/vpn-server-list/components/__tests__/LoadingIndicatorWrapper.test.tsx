import {render} from '@testing-library/react-native';
import React from 'react';
import LoadingIndicatorWrapper from '../LoadingIndicatorWrapper';
import {
  withReanimatedTimer,
  advanceAnimationByTime,
} from 'react-native-reanimated/lib/reanimated2/jestUtils';

describe('LoadingIndicatorWrapper', () => {
  test('render with isFetching = true', () => {
    withReanimatedTimer(() => {
      const tree = render(
        <LoadingIndicatorWrapper isFetching={true} />,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  test('render with isFetching = false', () => {
    withReanimatedTimer(() => {
      const tree = render(
        <LoadingIndicatorWrapper isFetching={false} />,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  test('given isFetching = true, container opacity is 1', () => {
    withReanimatedTimer(() => {
      const {queryByTestId} = render(
        <LoadingIndicatorWrapper isFetching={true} />,
      );

      const containerElement = queryByTestId(
        'vpn-server-list.loading-indicator.container',
      );

      expect(containerElement).toHaveAnimatedStyle({
        opacity: 1,
      });
    });
  });

  test('given isFetching = false, container opacity is 0', () => {
    withReanimatedTimer(() => {
      const {queryByTestId} = render(
        <LoadingIndicatorWrapper isFetching={false} />,
      );

      const containerElement = queryByTestId(
        'vpn-server-list.loading-indicator.container',
      );

      expect(containerElement).toHaveAnimatedStyle({
        opacity: 0,
      });
    });
  });

  test('given isFetching = true, when update isFetching to false, container will animate opacity from 1 to 0', () => {
    withReanimatedTimer(() => {
      const {queryByTestId, update} = render(
        <LoadingIndicatorWrapper isFetching={true} />,
      );

      const containerElement = queryByTestId(
        'vpn-server-list.loading-indicator.container',
      );

      expect(containerElement).toHaveAnimatedStyle({
        opacity: 1,
      });

      update(<LoadingIndicatorWrapper isFetching={false} />);

      // Default withTiming duration
      advanceAnimationByTime(300);

      expect(containerElement).toHaveAnimatedStyle({
        opacity: 0,
      });
    });
  });
});
