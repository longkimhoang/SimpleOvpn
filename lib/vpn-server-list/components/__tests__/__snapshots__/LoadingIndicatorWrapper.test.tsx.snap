// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicatorWrapper render with isFetching = false 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "opacity": 0,
      },
    }
  }
  collapsable={false}
  pointerEvents="none"
  style={
    Object {
      "height": "100%",
      "justifyContent": "center",
      "opacity": 0,
      "position": "absolute",
      "width": "100%",
    }
  }
  testID="vpn-server-list.loading-indicator.container"
>
  <View
    accessibilityRole="progressbar"
    accessibilityState={
      Object {
        "busy": false,
      }
    }
    accessible={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "height": 24,
          "opacity": 1,
          "width": 24,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "rgba(103, 80, 164, 1)",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "top": 12,
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": -12,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "rgba(103, 80, 164, 1)",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`LoadingIndicatorWrapper render with isFetching = true 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "opacity": 1,
      },
    }
  }
  collapsable={false}
  pointerEvents="none"
  style={
    Object {
      "height": "100%",
      "justifyContent": "center",
      "opacity": 1,
      "position": "absolute",
      "width": "100%",
    }
  }
  testID="vpn-server-list.loading-indicator.container"
>
  <View
    accessibilityRole="progressbar"
    accessibilityState={
      Object {
        "busy": true,
      }
    }
    accessible={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "height": 24,
          "opacity": 1,
          "width": 24,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "rgba(103, 80, 164, 1)",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "top": 12,
                "width": 24,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": -12,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "borderColor": "rgba(103, 80, 164, 1)",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
